### PR TITLE
[RFC][DNM] Asynchronous i2c and sensor API with hts221 example

### PR DESCRIPTION
**Pull request by nordic-krch**
It is a demonstration of proposed i2c and sensor API extension (#21538).

Following things are implemented:
- ~~`i2c_mngr` api and implementation~~
- ~~`i2c_ll api` and implementation for nordic platform.~~
- ~~extended `i2c` api to be able to use `i2c_mngr` if present and provide sync api on top of that
extended sensor API with async call~~
- modified `hts221` sensor to support reading temperature and humidity asynchronously
- modified `samples/sensor/hts221` to work in 2 modes: periodic sensor read from thread or asynchronously from k_timer (interrupt context). I've removed all printing to only get sensor related processing. Sample runs on nordic thingy52 (pca20020).~~

EDIT:
Changes:
- added `async_client` is a generic way of controlling asynchronous operations (extracted from `onoff` service)
- added draft of  `qop_mngr` module which is a generic module for performing queued operations. `qop_mngr` is using `async_client`
- added `i2c_common.c` which implements i2c_mngr which is capable of queueing i2c transactions.
- extened `i2c` API:
  - added `i2c_single_transfer` function for asynchronous single i2c message handling. This function must be implemented to enable asynchronous API for the driver.
  - added `struct i2c_transaction`
  - added helper macros for creating typical i2c transfers (`I2C_MSG_TXRX()`, `I2C_MSG_TXTX()`)
  - added `struct i2c_common_data` which contains `i2c_mngr` data and must be part of the driver data
  - added `i2c_schedule(dev, transaction)` for performing asynchronous transactions
- modified `i2c_nrfx_twim.c` which implements single transfer function
- modified HTS221 driver to use transaction for initialization

### CODE REVIEW

File: samples/sensor/hts221/src/main.c [hunk 1]
```diff
@@ -9,49 +9,105 @@
...
+
+K_TIMER_DEFINE(timer, expired, NULL);
+
 static void hts221_handler(struct device *dev,
 			   struct sensor_trigger *trig)
 {
 	process_sample(dev);
 }
 
+#ifdef CONFIG_PPI_TRACE
```
Comment (mbolivar) [inline samples/sensor/hts221/src/main.c]: Any pointers (or a Kconfig fragment) for how to run with this enabled on nrf52_pca20020 (which is how I assume you are testing this)? I tried running this sample from this branch inside NCS with this config overlay, but I never seem to get to main():

```
CONFIG_USE_SEGGER_RTT=y

# CONFIG_SHELL_BACKEND_RTT is not set

CONFIG_LOG=y
# CONFIG_LOG_BACKEND_SHOW_COLOR is not set
CONFIG_LOG_BACKEND_RTT=y

# CONFIG_UART_CONSOLE is not set
CONFIG_RTT_CONSOLE=y


CONFIG_PPI_TRACE=y

```

Comment (mbolivar) [reply samples/sensor/hts221/src/main.c]: > Any pointers (or a Kconfig fragment) for how to run with this enabled on nrf52_pca20020 (which is how I assume you are testing this)?

@nordic-krch ping?

Comment (nordic-krch) [reply samples/sensor/hts221/src/main.c]: @mbolivar sorry, got involved in something else. Will look at it tomorrow. Maybe I forgot to commit something.

Comment (mbolivar) [reply samples/sensor/hts221/src/main.c]: No problem, I will keep trying to figure it out for myself. Thanks.

Comment (nordic-krch) [reply samples/sensor/hts221/src/main.c]: i tried to run it and on my setup it also does not work well. After enabling logging i see that k_timer setup in main.c is constantly expiring even though it is set to expire 1sec. Unfortunately, i ended up on sick leave. Will get back into that once i get better.

Comment (mbolivar) [reply samples/sensor/hts221/src/main.c]: OK, thanks, and I hope you get better soon!

File: drivers/i2c/CMakeLists.txt [hunk 2]
```diff
@@ -12,7 +12,13 @@ zephyr_library_sources_ifdef(CONFIG_I2C_XEC		i2c_mchp_xec.c)
 zephyr_library_sources_ifdef(CONFIG_I2C_MCUX		i2c_mcux.c)
 zephyr_library_sources_ifdef(CONFIG_I2C_MCUX_LPI2C	i2c_mcux_lpi2c.c)
 zephyr_library_sources_ifdef(CONFIG_NRFX_TWI		i2c_nrfx_twi.c)
+
+if(CONFIG_I2C_USE_MNGR)
+zephyr_library_sources_ifdef(CONFIG_NRFX_TWIM		i2c_ll_nrfx_twim.c)
+else()
 zephyr_library_sources_ifdef(CONFIG_NRFX_TWIM		i2c_nrfx_twim.c)
+endif()
```
Comment (tbursztyka) [inline drivers/i2c/CMakeLists.txt]: Are you sure you want to do that? Unconditionally enabling i2c_mngr without letting other part of the code using the regular API is probably not a good idea. You cannot be sure some part of the system is still using the regular API vs this one. 

Plus, wasn't it so your API was supposed to use the regular API without requiring any driver changes?

**Review by tbursztyka — CHANGES_REQUESTED** [review]

**Comment by nordic-krch** [comment]
@pabigot please check where i got with queued op manager (`qop_mngr`) and asynchronous client:
- https://github.com/nordic-krch/zephyr/blob/i2c_async/include/sys/async_client.h
- https://github.com/nordic-krch/zephyr/blob/i2c_async/include/sys/qop_mngr.h

This satisfies my needs for i2c and spi asynchronous operations.

**Comment by mbolivar** [comment]
@nordic-krch I guess you can close this, since you said in https://github.com/zephyrproject-rtos/zephyr/pull/23075#issuecomment-591256066 that that PR is a replacement for this one?

**Comment by nordic-krch** [comment]
Closing this one as updated approach will be using queued operation manager and asynchronous sequence manager #23075.

**Comment by nordic-krch** [comment]
> Closing this one as updated approach will be using queued operation manager and asynchronous sequence manager #23075.

See #23371

**Pull request closed without merging**